x/pubsub/types/pubsub.go

```diff
@@ -0,0 +1,272 @@
+
+func ValidateDomain(domain string) error {
+	if domain == "" {
+		return fmt.Errorf("empty domain")
+	}
+
+	if len(domain) > MaxDomainLength {
+		return fmt.Errorf("domain over max length of %d: %d", MaxDomainLength, len(domain))
+	}
+
+	// TODO(bolten): perhaps we should add a regex to ensure a limited character set for the domain
+	if _, err := url.Parse(fmt.Sprintf("https://%s", domain)); err != nil {
+		return fmt.Errorf("invalid URL format: %s", err.Error())
+	}
+
+	return nil
+}
+
+func ValidateAddress(address string) error {
+	if address == "" {
+		return fmt.Errorf("empty address")
+	}
+
+	if _, err := sdk.AccAddressFromBech32(address); err != nil {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, fmt.Sprintf("invalid address: %s", err.Error()))
+	}
+
+	return nil
+}
+
+func ValidateProofUrl(proofUrl string, domain string, address string) error {
+	if proofUrl == "" {
+		return fmt.Errorf("empty proof URL")
+	}
+
+	if err := ValidateGenericUrl(proofUrl); err != nil {
+		return err
+	}
+
+	validProofUrl := fmt.Sprintf("https://%s/%s/cacert.pem", domain, address)
+	if proofUrl != validProofUrl {
+		return fmt.Errorf("invalid proof URL format, should be: %s", validProofUrl)
+	}
+
+	return nil
+}
+
+func ValidateGenericUrl(urlString string) error {
+	if urlString == "" {
+		return fmt.Errorf("empty URL")
+	}
+
+	if len(urlString) > MaxUrlLength {
+		return fmt.Errorf("URL over max length of %d: %d", MaxUrlLength, len(urlString))
+	}
+
+	if _, err := url.Parse(urlString); err != nil {
+		return fmt.Errorf("invalid URL format: %s", err.Error())
+	}
+
+	return nil
+}
+
+func ValidateCaCertificateBase64(certBase64 string) error {
+	if certBase64 == "" {
+		return fmt.Errorf("empty CA certificate")
+	}
+
+	if len(certBase64) > MaxCertLength {
+		return fmt.Errorf("CA cert over max length of %d: %d", MaxCertLength, len(certBase64))
+	}
+
+	certBytes, err := base64.StdEncoding.DecodeString(certBase64)
+	if err != nil {
+		return fmt.Errorf("invalid base64 encoding for CA cert")
+	}
+
+	block, rest := pem.Decode(certBytes)
+	if block == nil || len(rest) > 0 {
+		return fmt.Errorf("invalid PEM formating, expecting only certificate block")
+	}
+
+	if block.Type != "CERTIFICATE" {
+		return fmt.Errorf("invalid PEM certificate block, must be CERTIFICATE")
+	}
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return fmt.Errorf("could not parse x509 certificate")
+	}
+
+	if !cert.IsCA {
+		return fmt.Errorf("must be a CA cert")
+	}
+
+	if cert.PublicKeyAlgorithm != x509.ECDSA {
+		return fmt.Errorf("must be an ECDSA cert")
+	}
+
+	return nil
+}
+
+func ValidateSubscriptionId(subscriptionId string) error {
+	if subscriptionId == "" {
+		return fmt.Errorf("empty subscription ID")
+	}
+
+	if len(subscriptionId) > MaxSubscriptionIdLength {
+		return fmt.Errorf("subscription ID over max length of %d: %d", MaxSubscriptionIdLength, len(subscriptionId))
+	}
+
+	// TODO(bolten): any other character limitations we should add here?
```

Comment:
Maybe safer to say what is allowed via regex rather than what isn't? And then allow it to be configured since this is a generic module.

```
/^[A-z0-9-_]+$/
```

Reply:
This is a good point, a lot of constants and other choices can probably be in Params instead, which is currently empty.

Reply:
To your other point, I think the point of specifying only invalid characters was to maximize flexibility for the format of a subscription ID. The pipe is only blocked because it's the keyspace delimiter.